### ISSUE TITLE
bi vs 2bi and stack heights

### ISSUE THREAD

**Issue by catb0t** [factor/factor#1939]
I have never thought that `[ ] [ ] bi` would be the apparent source of a bug for me, but here we are. 

Here is a compressed version of the code in question. It is from 

https://github.com/catb0t/factor/blob/a7ba6cfee56a1aa13fba77d134cb13b8d3a86fa4/basis/math/matrices/matrices.factor#L349-L355

We are only interested in 2 lines from this, but it is minimally recreated here so you can run it without cloning my fork or downloading the entire file. 

```factor
USING: arrays fry kernel math math.matrices math.vectors sequences ;

: (1determinant) ( matrix -- 1det ) 1 ;
: (2determinant) ( matrix -- 2det ) 2 ;
: (3determinant) ( matrix-seq -- 3det ) 3 ;

: cols-except ( matrix n -- others )
    [ dup dim second <iota> ] dip
    '[ _ = ] { } reject-as swap cols flip ;

: alternating-sign ( seq odd-elts? -- seq' )
    '[ 2 mod zero? _ = [ neg ] unless ] map-index ;

DEFER: (determinant)
DEFER: (ndeterminant)
: make-determinants ( n mat -- seq )
    <repetition> [
        cols-except [ length ] keep (ndeterminant) ! recurses here
    ] map-index ;

: (ndeterminant) ( n mat -- ndet )
    over 3 < [ (determinant) ] [
        ! [ first t alternating-sign ] [ rest make-determinants ] bi ! === significant line
        dup first t alternating-sign -rot rest make-determinants ! === significant line
        v* sum
    ] if ;

! switches on dimensions only
: (determinant) ( n: integer matrix: matrix -- determinant ) over {
        { 1 [ nip (1determinant) ] }
        { 2 [ nip (2determinant) ] }
        { 3 [ nip (3determinant) ] }
        [ drop (ndeterminant) ]
    } case ; 
```

The lines have the same effect and semantics when there are two items on the stack.

```factor
[ first t alternating-sign ] [ rest make-determinants ] bi
dup first t alternating-sign -rot rest make-determinants
```
The first line can be written as its `keep` equivalent for the exact same semantics and effect as `bi` too.

**Note as well the mutual recursive nature of `(ndeterminant)` and `make-determinants`.**

The point of these two lines is to
1. call `t alternating-sign` on the first element of the 2D sequence (matrix) on the top of the stack, leaving a sequence 
2. call `make-determinants` with `n` and the matrix except its first element (`rest` or `1 tail`) 

Once these lines have executed, the stack should be, for example (with topmost stack item lowest) 
```factor
{ 1 -2 3 -4 } ! from alternating-sign
{ 44 -6 11 12 } ! from make-determinants 
```
So far, so good. What is the matter? 

The problem appears to be that when using the `[ ] [ ] bi` (or `[ ] keep`) line, the second quotation operates on a reference to the input matrix of the first quotation. 

The consistent error message gives this impression. 

```factor
Unit Test: {
    { t }
    [
        {
            { -5 4 -3 2 }
            { -2 1 0 -1 }
            { -2 -3 -4 -5 }
            { 0 2 0 4 }
        } {
           4 swap (ndeterminant)
    ]
}

No suitable arithmetic method
left    { -5 -4 -3 -2 }
right   0
generic <
```
The error is because the lower stack operand to `make-determinants` (what should have been an integer `n`) was instead `{ -5 -4 -3 -2 }` -- this is the output of `alternating-sign`.

Pay specific attention to the signs of the numbers in the first element of the input matrix -- this is what shows that the output of the first quotation in the `bi` is taken into the second quotation. 
 
In the original matrix, the first row is `{ -5 4 -3 2 }`. Indices `0` and `2` are already signed, but indices `1` and `3` are not. 

The error contains the row `{ -5 -4 -3 -2 }`, in which all indices `0 1 2 3` are signed. `alternating-sign` changes the sign of odd indices when given `t` as an input.

This way, `{ -5 -4 -3 -2 }` is in two places at once. Seemingly because of the `bi`, it is simultaneously  
* the output of the first quotation; the "mutated" first element in `mat` (correctly)
* ***and*** the leftmost input to `make-determinants`, as if it was on the bottom of the stack when `rest` was called (incorrectly)

I came upon this conclusion through liberal application of `TYPED:`. 

But when the equivalent (yet still recursive) code is written using shuffle words (`dup` and `-rot`) rather than a `bi` or a `keep`, the code works properly 100% of the time, and `make-determinants` does not get any incorrect arguments.

But if you write in the listener: 
```factor
{
            { -5 4 -3 2 }
            { -2 1 0 -1 }
            { -2 -3 -4 -5 }
            { 0 2 0 4 }
}
[ first t alternating-sign ] [ rest ] bi . .
```
Everyone knows what the output will be: 
```factor
{ { -2 1 0 -1 } { -2 -3 -4 -5 } { 0 2 0 4 } }
{ -5 -4 -3 -2 }
```

This is the output, because `array`s are not mutable, and because `first` (`nth`) can't mutate its input either. 

How has this happened? Is there a bizarre bug in Factor, or (more likely) have I written a bug myself? Or is it a problem with the bi-recursive words operating on sequences?

**Comment by bjourne** [comment]
Your report is quite hard to follow. But the lines
```factor
[ first t alternating-sign ] [ rest make-determinants ] bi
dup first t alternating-sign -rot rest make-determinants
```
are not equivalent which you can see if you wrap them in `infer.` like this `[ code for line 1 ] infer.`

**Comment by catb0t** [comment]
The stack effects are not identical, but they both leave the stack having the same height it did previously. `( x -- x )` vs `( x x -- x x )`; the word with stack effect `( n mat -- ndet )` compiles either way.

Sorry the explanation is so poor, I can try to rewrite it. I was attempting to recreate my thought process when understanding the reason for the problem, which I had only recently discovered after a lot of testing.

**Comment by catb0t** [comment]
Oh, it's a rather simple problem I think. The input quotations to `bi` in my code (obviously) do not have the same stack effects.

`[ first t alternating-sign ]` has the effect of `( x -- x )`
`[ rest swap . ]` has the effect of `( x x -- x )`, same as the code above

The line 
```factor
9 { { 1 2 3 } { 4 5 6 } } [ first t alternating-sign ] [ rest swap . ] bi
```
makes it very clear what is going on here -- because the input quotations to `bi` have different input stack heights, the input to the second quotation is the output of the first quotation.

`{ 1 -2 3 }` will be printed and `9` will be left on the stack below the result of `rest`.

The bug boils down to my own ignorance toward `bi`'s specific behaviour. 

Therefore I think that either the input quotations to `bi` should either: 
* be required to have the same input stack height, or equivalent stack effects altogether 
* at least be documented to have this (poorly-defined? unintended?) tricky behaviour

**Comment by bjourne** [comment]
`bi`s with quotations with different stack effects are occasionally useful. F.e `2 [ sq ] [ ] bi` They are documented in the `dataflow-combinators` article in the help. You are right that the help pages could be improved here.
